File: Python/src/item_factory.py
from enum import Enum

from src.item import Item
from src.models.aged_brie import AgedBrie
from src.models.backstage_pass import BackstagePass
from src.models.conjured_item import ConjuredItem
from src.models.specialized_item import SpecializedItem
from src.models.standard_item import StandardItem
from src.models.sulfuras import Sulfuras


class ItemType(Enum):
    BRIE = "Aged Brie"
    BACKSTAGE_PASSES = "Backstage passes to a TAFKAL80ETC concert"
    SULFURAS = "Sulfuras, Hand of Ragnaros"
    CONJURED_ITEM = "Conjured"


def get_specific_item_from_name(item: Item) -> SpecializedItem:
    match item.name:
        case ItemType.BRIE:
            return AgedBrie(item)
        case ItemType.BACKSTAGE_PASSES:
            return BackstagePass(item)
        case ItemType.SULFURAS:
            return Sulfuras(item)
        case ItemType.CONJURED_ITEM:
            return ConjuredItem(item)
        case _:
            return StandardItem(item)

File: Java/src/main/java/com/gildedrose/ItemFactory.java
package com.gildedrose;

import com.gildedrose.items.*;

import java.util.HashMap;
import java.util.Map;

public class ItemFactory {
    public final static String BRIE = "Aged Brie";
    public final static String BACKSTAGE_PASSES = "Backstage passes to a TAFKAL80ETC concert";
    public final static String SULFURAS = "Sulfuras, Hand of Ragnaros";
    public final static String CONJURED_ITEM = "Conjured";
    private final static Map<String, SpecializedItem> itemsMap = new HashMap<>();

    public ItemFactory(Item item) {
        itemsMap.put(BRIE, new AgedBrie(item));
        itemsMap.put(BRIE, new AgedBrie(item));
        itemsMap.put(BACKSTAGE_PASSES, new BackstagePasses(item));
        itemsMap.put(SULFURAS, new Sulfuras(item));
        itemsMap.put(CONJURED_ITEM, new ConjuredItem(item));
    }

    public SpecializedItem getSpecificItemFromName(Item item) {
        if (isStandardItem(item))
            return new StandardItem(item);
        return itemsMap.get(item.name);
    }

    private boolean isStandardItem(Item item) {
        return !itemsMap.keySet().contains(item.name);
    }
}
